lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+    end
+
+    -- Figure out the maximum size we can give out to sub-widgets
+    local sub_width = max_value(cols_size)
+    local sub_height = max_value(rows_size)
+    if self._private.horizontal_expand then
+        sub_width  = (width - 2*hpadding - (self._private.num_cols-1)*hspacing) / self._private.num_cols
+    end
+    if self._private.vertical_expand then
+        sub_height = (height - 2*vpadding - (self._private.num_rows-1)*vspacing) / self._private.num_rows
+    end
+
+    -- Calculate the position and size to place the widgets
+    local cumul_width, cumul_height = {}, {}
+    local cw, ch = hpadding, vpadding
+    for j = 1, #cols_size do
+        cumul_width[j] = cw
+        if self._private.horizontal_fixed_size then
+            cols_size[j] = math.max(self._private.min_cols_size, sub_width)
+        elseif self._private.horizontal_expand then
+            local hpercent = self._private.num_cols * sub_width * cols_size[j] / total_expected_width
+            cols_size[j] = math.max(self._private.min_cols_size, hpercent)
+        end
+        cw = cw + cols_size[j] + hspacing
+    end
+    cumul_width[#cols_size + 1] = cw
+    for i = 1, #rows_size do
+        cumul_height[i] = ch
+        if self._private.vertical_fixed_size then
+            rows_size[i] = math.max(self._private.min_rows_size, sub_height)
+        elseif self._private.vertical_expand then
+            local vpercent = self._private.num_rows * sub_height * rows_size[i] / total_expected_height
+            rows_size[i] = math.max(self._private.min_rows_size, vpercent)
+        end
+        ch = ch + rows_size[i] + vspacing
+    end
+    cumul_height[#rows_size + 1] = ch
+
+    -- Place widgets
+    local fill_space = true  -- should be fill_space property?
+    for _, v in pairs(self._private.widgets) do
+        local x, y, w, h
+        -- Round numbers to avoid decimals error, force to place tight widgets
+        -- and avoid redraw glitches
+        x = math.floor(cumul_width[v.col])
+        y = math.floor(cumul_height[v.row])
+        w = math.floor(cumul_width[v.col + v.col_span] - hspacing - x)
+        h = math.floor(cumul_height[v.row + v.row_span] - vspacing - y)
+        -- Recalculate the width so the last widget fits
+        if (fill_space or self._private.horizontal_expand) and x + w > math.floor(width - hpadding) then
+            w = math.floor(math.max(self._private.min_cols_size, width - hpadding - x))
+        end
+        -- Recalculate the height so the last widget fits
+        if (fill_space or self._private.vertical_expand) and y + h > math.floor(height - vpadding) then
+            h = math.floor(math.max(self._private.min_rows_size, height - vpadding - y))
+        end
+        -- Place the widget if it fits in the area
+        if x + w <= math.floor(width - hpadding) and y + h <= math.floor(height - vpadding) then
+            table.insert(result, base.place_widget_at(v.widget, x, y, w, h))
+        end
+    end
```

Comment:
I know that `:layout` and `:fit` are always quite harry, but still I barely understand this one. Somehow it feels to me like this has too many options, or something like that.